### PR TITLE
GH2487: Add alias dupe check on method/property signarture

### LINKED ISSUE #2487: Warn and skip code gen for duplicate aliases
PR #2455 created a conflict with existing Cake.Incubator addin
```diff
- (2688,10): error CS0111: Type 'Submission#0' already defines a member called 'EnvironmentVariable' with the same parameter types
```

Repro script
```cake
#addin nuget:?package=Cake.Incubator&version=3.1.0
```

What we think we need to do here is track methods generated by code gen and skip & warn on any duplicates, prioritizing what's in Cake.Common.

Likely would need to add using statements to avoid errors with aliases being invoked on `ICakeContext`

### PR DESCRIPTION
**Pull request by devlead**
Fixes #2487

### CODE REVIEW

File: src/Cake/Scripting/Roslyn/RoslynCodeGenerator.cs [hunk 1]
```diff
@@ -22,14 +22,38 @@ public string Generate(Script script)
...
         {
-            var result = new List<string>();
+            var result = new Dictionary<string, string>();
             foreach (var alias in context.Aliases)
             {
-                result.Add(alias.Type == ScriptAliasType.Method
-                    ? MethodAliasGenerator.Generate(alias.Method)
-                    : PropertyAliasGenerator.Generate(alias.Method));
+                string hash,
+                       code = alias.Type == ScriptAliasType.Method
```
Comment (patriksvensson) [inline src/Cake/Scripting/Roslyn/RoslynCodeGenerator.cs]: Can we put this on the same line as the one above?

File: src/Cake.Core/Scripting/CodeGen/PropertyAliasGenerator.cs [hunk 2]
```diff
@@ -150,7 +160,25 @@ private static string GenerateCode(MethodInfo method)
             return builder.ToString();
         }
 
-        private static string GenerateCachedCode(MethodInfo method)
+        private static string GenerateCommonInitalCode(MethodInfo method, StringBuilder builder)
```
Comment (patriksvensson) [inline src/Cake.Core/Scripting/CodeGen/PropertyAliasGenerator.cs]: Can we make pass `StringBuilder` using `ref`? Not that it matters but indicates to the reader that we're mutating it inside the method.

File: src/Cake.Core/Scripting/Script.cs [hunk 3]
```diff
@@ -74,12 +72,13 @@ public sealed class Script
...
-            _usingAliasDirectives = new List<string>(usingAliasDirectives ?? Enumerable.Empty<string>());
-            _usingStaticDirectives = new List<string>(usingStaticDirectives ?? Enumerable.Empty<string>());
-            _defines = new List<string>(defines ?? Enumerable.Empty<string>());
+            Namespaces = new List<string>(namespaces ?? Enumerable.Empty<string>());
+            Lines = new List<string>(lines ?? Enumerable.Empty<string>());
+            Aliases = new List<ScriptAlias>(aliases ?? Enumerable.Empty<ScriptAlias>());
+            UsingAliasDirectives = new List<string>(usingAliasDirectives ?? Enumerable.Empty<string>());
+            UsingStaticDirectives = new List<string>(usingStaticDirectives ?? Enumerable.Empty<string>());
+            Defines = new List<string>(defines ?? Enumerable.Empty<string>());
+            ExcludedNamespaces = new Dictionary<string, IList<string>>(System.StringComparer.Ordinal);
```
Comment (patriksvensson) [inline src/Cake.Core/Scripting/Script.cs]: Do we need to fully qualify the namespace here?

File: src/Cake/Scripting/Roslyn/RoslynScriptSession.cs [hunk 4]
```diff
@@ -82,9 +82,17 @@ public void Execute(Script script)
             var generator = new RoslynCodeGenerator();
             var code = generator.Generate(script);
 
+            // Warn about any code generation excluded namespaces
+            foreach (var @namespace in script.ExcludedNamespaces)
+            {
+                _log.Warning("Namespace {0} excluded by code generation, affected methods:\r\n\t{1}",
+                @namespace.Key,
```
Comment (patriksvensson) [inline src/Cake/Scripting/Roslyn/RoslynScriptSession.cs]: Can we indent this line and move the line under this up?

**Comment by devlead** [comment]
@patriksvensson feedback addressed, removed WIP.

File: src/Cake/Scripting/Roslyn/RoslynCodeGenerator.cs [hunk 5]
```diff
@@ -22,14 +22,37 @@ public string Generate(Script script)
...
-                    ? MethodAliasGenerator.Generate(alias.Method)
-                    : PropertyAliasGenerator.Generate(alias.Method));
+                string hash, code = alias.Type == ScriptAliasType.Method
+                    ? MethodAliasGenerator.Generate(alias.Method, out hash)
+                    : PropertyAliasGenerator.Generate(alias.Method, out hash);
+
+                string @namespace = alias.Method.DeclaringType.Namespace ?? "@null";
+                if (result.ContainsKey(hash))
+                {
+                    var message = $"{alias.Type} \"{alias.Name}\" excluded from code generation and will need to be fully qualified on ICakeContext.";
```
Comment (patriksvensson) [inline src/Cake/Scripting/Roslyn/RoslynCodeGenerator.cs]: This is nitpicking but could we change this to

`The {alias.Type.ToLowerInvariant()} \"{alias.Name}\" was excluded from code generation and will need to be fully qualified on ICakeContext.`

Comment (devlead) [reply src/Cake/Scripting/Roslyn/RoslynCodeGenerator.cs]: Addressed.

**Comment by patriksvensson** [comment]
Merged! Thank you for your contribution. Much appreciated! 👍

**Pull request merged**